File: c++/0938.cpp
// Link: https://leetcode.com/problems/range-sum-of-bst/description/
// Runtime: 98 ms 
// Memory: 64.95 MB

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution 
{
public:
    int rangeSumBST(TreeNode* root, int low, int high) 
    {
        if (root == nullptr)
            return 0;
        else
        {
            int currentSum = 0;
            if ((root->val >= low) && (root->val <= high))
                currentSum += root->val;
            
            return currentSum + rangeSumBST(root->left, low, high) + rangeSumBST(root->right, low, high);
        }
    }
};

File: python/0938.py
# Link: https://leetcode.com/problems/range-sum-of-bst/description/
# Runtime: 120 ms
# Memory: 24.41 MB

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rangeSumBST(self, root: Optional[TreeNode], low: int, high: int) -> int:
        if root:
            current_sum = 0
            if low <= root.val <= high:
                current_sum += root.val

            return (
                current_sum
                + self.rangeSumBST(root.left, low, high)
                + self.rangeSumBST(root.right, low, high)
            )
        else:
            return 0
